### DISCUSSION TITLE
[0.14.0 Beta 2]: Proxy auth not working

### DISCUSSION THREAD

**Discussion by miguelangel-nubla** [blakeblackshear/frigate#11946] [Report a Bug] [answered]
### Describe the problem you are having

It works fine with `native` auth, but as soon as I set the mode to `proxy`, Frigate behaves as if no auth is configured.
It shows `Current User: anonymous` on the UI and nothing in logs about auth or any error.

### Steps to reproduce

In a working native auth installation change the auth mode to proxy.

### Version

0.14.0-bccffe6

### Frigate config file

```yaml
auth:
  mode: proxy
  header_map:
    user: x-forwarded-user
```

### Relevant log output

```shell
Nothing relevant even with debug logging.
```

### Operating system

Other

### Install method

Docker CLI

### Network connection

Wired

### Camera make and model

NA

**Comment by NickM-27** [comment, marked as answer]
> It works fine with native auth, but as soon as I set the mode to proxy, Frigate behaves as if no auth is configured.

that is exactly how it is supposed to work. Proxy auth means your proxy is providing the auth

**Reply by miguelangel-nubla** [reply]
The header provided by the proxy is not checked at all.
I expected it to block requests that do not send a valid header with a valid user form a trusted proxy.

What is the point if that is not the case? I mean I have been using it with a reverse proxy and external auth forever, what does the new auth proxy mode adds then?

**Reply by NickM-27** [reply]
the proxy mode adds support for commonly used auth headers so frigate will be aware of the username so when we add user roles it will translate over directly

**Reply by NickM-27** [reply]
this is explained in the docs https://deploy-preview-11419--frigate-docs.netlify.app/configuration/authentication#proxy-mode

**Reply by miguelangel-nubla** [reply]
I understand, but for that to be of any use invalid requests must be denied.

I get that user roles is a _planned_ feature, but even with that out of the picture proxy mode should deny invalid requests.
Otherwise it does exactly the same as not setting `auth:` on the config, and I think that should be made clear in the docs since that is not the expected way proxy auth works.

**Reply by blakeblackshear** [reply]
Do you have any examples of how other solutions identify "invalid requests" that isn't bypassed by simply passing a header.

**Reply by miguelangel-nubla** [reply]
Of the top of my mind these two are a must:
- Requests coming from a IP not specified as `trusted_proxy`.
- Requests coming from a `trusted_proxy` that do not have all headers configured in `header_map` (user, group (in the future), etc).

And additionally:
- Requests coming from a `trusted_proxy` that do have valid headers but that user/group does not exist.
This depends on the future implementation for RBAC.

  - If all users are "known" (in the database) it is trivial to check.
This way has the advantage that since the username is forced by the proxy and not modifiable by the end user, it enables a deny by default behavior. Newly created users in the proxy (or in the common case proxy takes existing users from a global database) will be denied access until created in Frigate.

  - If the username is allowed to be "new"/unknown-yet-to-frigate, in order to allow default user roles on existing proxy users (which I think is not a good fit for this particular use case) this one would not apply.

**Reply by blakeblackshear** [reply]
> Requests coming from a IP not specified as trusted_proxy.

This could be implemented, but wouldn't work for the vast majority of users running the proxy in docker since the docker IP won't be consistent. It would be easy to add, but would only be useful for a very small number of users.

> Requests coming from a trusted_proxy that do not have all headers configured in header_map (user, group (in the future), etc).

This is not an actual security measure and has no value as those headers are trivial to figure out.

> Requests coming from a trusted_proxy that do have valid headers but that user/group does not exist.
This depends on the future implementation for RBAC.

All of these are also trivial to bypass and offer no real security value.

The only way to actually do this properly would be to have the upstream proxy sign the request with with a known secret and validate the signature matches in Frigate. This is how payment platforms like stripe authenticate webhooks. I would have to look and see if any of the commonly used proxy servers support something like this.

**Reply by miguelangel-nubla** [reply]
> > Requests coming from a IP not specified as trusted_proxy.
> 
> This could be implemented, but wouldn't work for the vast majority of users running the proxy in docker since the docker IP won't be consistent. It would be easy to add, but would only be useful for a very small number of users.

That depends on the network architecture each one decides to run on.
It is absolutely possible to do it right in docker using docker networks and IP ranges.
Of course a reverse proxy with proper auth forwarding is not some one click task, and therefore useful for not many users, but it makes a world of difference to us, the minority, in the absence of alternatives.

In my opinion, the majority will be much better served with a standard reverse proxy and `native` frigate auth.

> > Requests coming from a trusted_proxy that do not have all headers configured in header_map (user, group (in the future), etc).
> 
> This is not an actual security measure and has no value as those headers are trivial to figure out.

If your host where the proxy resides also runs something that can make simple http calls to arbitrary URLs (which is very common), this prevents it accessing Frigate. It has the same value as the last one (ARP poisoning enters the chat), only a basic check, but it can be quite useful for what easy it is to implement.

> All of these are also trivial to bypass and offer no real security value.

I agree.
I only mention these because I understood the devs were not interested on making it "secure", and were leaving that job to reverse proxies.



-------------

If instead the focus is on security done right, these are the options I can think of:

1. The compatible way with decent "security" would be to use regular TLS on the Frigate server and set a "secret" header on the reverse proxy such as `X-Secret-Token=h7WzrMEamuQ7qQfvTKZueHCfCIv` to be validated by Frigate.
This is compatible with every reverse proxy I can think of.
3. Mutual TLS, which is equivalent to what you described but it is already standardized.
You can do some fun stuff with certificate chains, but at the end of the day all comes down to a single "secret" inside Frigate like the first option.

**Reply by blakeblackshear** [reply]
I updated the docs to make it clear what proxy mode means. Going to close this as it's not a bug. If we get a lot of feedback from users then perhaps this will be implemented one day. I would also consider accepting a PR.

**Reply by miguelangel-nubla** [reply]
I am willing to create a PR for this.
Is a potential PR waiting on future user feedback or can we just agree on the path forward for the expected behavior?
I would personally go with the secret header over **optional** TLS to simplify it for everyone.

**Reply by blakeblackshear** [reply]
A shared secret header is fine. Users can decide if they want to use TLS along with it. No need to tangle those things together.

**Reply by miguelangel-nubla** [reply]
This is conceptually what I have in mind, simplifying and removing the explicit `proxy` mode.

Configuration:
```
auth:
  # Optional: Reset the admin user password on startup (default: shown below)
  # New password is printed in the logs
  reset_admin_password: False
  # Optional: Cookie to store the JWT token for native auth (default: shown below)
  cookie_name: frigate_token
  # Optional: Set secure flag on cookie. (default: shown below)
  # NOTE: This should be set to True if you are using TLS
  cookie_secure: False
  # Optional: Session length in seconds (default: shown below)
  session_length: 86400 # 24 hours
  # Optional: Refresh time in seconds (default: shown below)
  # When the session is going to expire in less time than this setting,
  # it will be refreshed back to the session_length.
  refresh_time: 43200 # 12 hours
  # Optional: Number of hashing iterations for user passwords
  # As of Feb 2023, OWASP recommends 600000 iterations for PBKDF2-SHA256
  # NOTE: changing this value will not automatically update password hashes, you
  #       will need to change each user password for it to apply
  hash_iterations: 600000
  # Optional: Rate limiting for login failures to help prevent brute force
  # login attacks (default: shown below)
  # See the docs for more information on valid values
  failed_login_rate_limit: None
  # Optional: Allowed trusted proxies.
  reverse_proxy:
    - edge-proxy:
        secret: long-secure-private-secret1
        # Mapping of headers received from the proxy (default: shown below)
        header_map:
          # Used for rate limiting.
          client: x-forwarded-for
          # If no valid user header is received the request will be handled by regular auth.
          user: x-forwarded-user
        # Optional: URL where to redirect the user for logging out in the
        # downstream reverse proxy.
        # Only for clients correctly authenticated by headers (default: shown below)
        logout_url: /api/logout
```
Requests to the authenticated endpoint 8080 will always be handled with native authentication, except under the following conditions:

- The `X-Secret-Token` header matches the token specified in `auth.reverse_proxy.[name].secret`.
- An additional `user` header is received, as configured, and is valid.

These conditions override the native authentication. Invalid requests will receive a `401 Unauthorized` HTTP response code and will be redirected to the login page to allow simpler use cases to work.

**Reply by blakeblackshear** [reply]
The vast majority of users aren't going to care about this setting because their reverse proxy sits on the same internal docker network as frigate. This is what I had in mind.

You still need a `mode` because not all proxies support passing a user down in a header. You can't assume you need to enforce native auth in that situation.

```yaml
auth:
  mode: proxy
  proxy:
    authentication_method: None # defaults to None and available options are None and shared_secret
    shared_secret: <secret_value> # it needs to be possible to pass this with a docker secret or env var
    header_map:
      user: x-forwarded-user
    logout_url: /api/logout
```

Rate limiting is only implemented on the login url, and I don't want to make changes to that along with this.

I can't say I have personally seen other self hosted projects implement this pattern. If you have seen them, I would appreciate you listing some before it's implemented so we can check the implementation against others.

**Reply by miguelangel-nubla** [reply]
> The vast majority of users aren't going to care about this setting because their reverse proxy sits on the same internal docker network as frigate. This is what I had in mind.

Exactly, I am one of those users, and I don't want other services in my network to be able to access/intercept frigate connections.
That is the whole point of all this reverse-proxy to upstream server security.

Also what about Home Assistant users? Those are very likely to *not* be on the same machine. This enables the frigate proxy addon to solve that use case with just a header, with all the options to use/not use user based access for those who want it.

I just want a Frigate instance that I cannot query from anywhere (internal or external network) but the authorized services.
I want to decide who sees my cameras, even on the internal network.

> You still need a `mode` because not all proxies support passing a user down in a header.

No, if you want native auth, point the reverse proxy to 8080. With or without another external proxy auth. If you don't need native auth you point it to 5000 just like until now. With or without another external proxy auth.
This covers every use case.
What is the point of having a unauthenticated port (5000) and a authenticated port (8080) if the auth port is *sometimes* not protected (proxy mode)? Why the redundancy and the extra complexity for no real benefit?


> Rate limiting is only implemented on the login URL, and I don't want to make changes to that along with this.

Sure, that is just fine. But you **need** the client IP for requests coming from reverse proxy. Otherwise everyone gets blocked.

> I can't say I have personally seen other self hosted projects implement this pattern. If you have seen them, I would appreciate you listing some before it's implemented so we can check the implementation against others.

I don't know exactly which part are you referring to with pattern. I will answer generally, let me know if I missed the point of the question.

If you are referring to passing a secret to a server in order to allow the server to identify the middleman, every service that requires you to use API keys usually does this. They just call it API key.

The usual way this is done for reverse proxies is with those simple "checks" we discussed earlier that "offer no real security". That is because it is expected to use them in conjunction with proper firewall rules and proper network separation.
I agree everything has to be done just right for those "checks" to be useful, but those implementations fully rely on those checks, rejecting everything else, and do not make sense without them.

Once you stated those checks are of "no real security value", the next step is encryption and mutual auth. Everyone that does this jump for these use cases goes directly to mTLS or similar because it can work with multiple peers, and be configured in a way that you don't have to touch every server to add/remove a single point, and instead rely on PKI. Examples:
- [NGINX](https://docs.nginx.com/nginx/admin-guide/security-controls/securing-http-traffic-upstream/) example. [Step-ca](https://smallstep.com/hello-mtls/doc/client/nginx-proxy) is great for easing things up.
- [Traefik version](https://smallstep.com/hello-mtls/doc/combined/traefik/nginx-proxy)

Doing it with a secret is the same pattern, but simplified in order to not require a PKI, which is way overkill for Frigate specially because Frigate instances are not expected to be scaled horizontally.

**Reply by blakeblackshear** [reply]
> No, if you want native auth, point the reverse proxy to 8080. With or without another external proxy auth. If you don't need native auth you point it to 5000 just like until now. With or without another external proxy auth.
This covers every use case.

It doesn't cover the use case where all I want is for Frigate to recognize my user so it can be displayed in the UI and ultimately used for RBAC. I don't want to configure a `secret_token`, because I'm not worried about other containers in my docker compose making roque queries to Frigate. I guess we could apply the header_map regardless of port before bypassing the remaining authentication logic since it doesn't actually have anything to do with authentication and that would result in a clean split where 8080 requires either a `secret_token` or a valid JWT token and 5000 doesn't do any auth checks.

> Sure, that is just fine. But you need the client IP for requests coming from reverse proxy. Otherwise everyone gets blocked.

You will find that `X-Forwarded-For` is already parsed and identifies the client ip by skipping trusted proxies. The header name just isn't customizable.

**Reply by miguelangel-nubla** [reply]
> > No, if you want native auth, point the reverse proxy to 8080. With or without another external proxy auth. If you don't need native auth you point it to 5000 just like until now. With or without another external proxy auth.
> > This covers every use case.
> 
> It doesn't cover the use case where all I want is for Frigate to recognize my user so it can be displayed in the UI and ultimately used for RBAC. I don't want to configure a `secret_token`, because I'm not worried about other containers in my docker compose making roque queries to Frigate.

So you are talking about a use case where you already have set a custom dynamic header for sending the user.
Just add another header with value `X-Secret-Token=hello` or whatever you want to call it. Done, use case covered.


> I guess we could apply the header_map regardless of port before bypassing the remaining authentication logic since it doesn't actually have anything to do with authentication

I don't understand why you would want to do that. Port 5000 does not do auth, and does not do proxy or header parsing, just like right now.
Port 8080 does both, taking the user from `header_map` if wanted, which is certainly authentication.

>  and that would result in a clean split where 8080 requires either a `secret_token` or a valid JWT token and 5000 doesn't do any auth checks.

You are mixing things up, I suggested port 8080 always requires auth (frigate knows the user), either from header from reverse proxy or native auth.
You can have another independent auth on the reverse proxy and not send the header if you wanted. Or just not have it.



> > Sure, that is just fine. But you need the client IP for requests coming from reverse proxy. Otherwise everyone gets blocked.
> 
> You will find that `X-Forwarded-For` is already parsed and identifies the client ip by skipping trusted proxies. The header name just isn't customizable.

That does not make any sense. You have to trust the header, or else in setups without reverse proxy any client can send its own "simulated IP" and the rate limit will not apply to it. That is the whole point of needing to trust proxies.
